### DISCUSSION TITLE
Questions regarding -limit threads and throttle

### DISCUSSION THREAD

**Discussion by AV-Gaming** [ImageMagick/ImageMagick#7720] [Help] [answered]
### ImageMagick version

ImageMagick 7.1.0-5 Q16

### Operating system, version and so on

Windows 

### Description

Hey guys,

Just some hopefully quick questions. If this is not the forum for asking questions about ImageMagick, then I apologise, but please let me know where I should ask this.

I am porting an ImageMagick script to Windows, from Bash, and everything is going well, but the command, when run, maxes out all available CPU threads, which is of course, totally normal. I wanted to add some optional parameters to my script to allow users to make the process less intensive.

I read up and found about the "threads" and "throttle" limit options.

I know the below works:

`magick -limit thread X` ...and then the rest of the script (where X is the number of threads)

And I know the below runs, but I can't tell if it makes a difference honestly, task manager still says 100%:

`magick -limit throttle 100` ...and then the rest of the script  (100 can be any number, really, within reason, in ms)

So the obvious question I had was about combining these two. The below doesn't work:

`magick -limit thread X throttle Y` ...and then the rest of the script

But the below does:

`magick -limit thread X -limit throttle Y` ...and then the rest of the script

So question 1) When someone wants to specify both a "throttle" and a "thread" number for a magick process, is that the right way to do it? Both specified at the start, one after the other, with a "-limit" before them? And if so, does it matter if thread or throttle goes first? Do both get respected either way?

That's question 1.

Really I would be happy to just add the ability to limit number of threads, but I was thinking about peeps with older computers who may only have one core, the thread thing would be useless in such an instance. So regarding throttle, this seems to allow "breathing room" by occasionally pausing, but can anyone say if this is actually a useful feature? Is there any point using it? I have 4 threads on my CPU, and I tried specifying 4 threads, and a throttle time of 100ms, then 300ms - task manager still reported 100% usage and my PC still started stuttering until the process was complete. Maybe I am giving it the wrong values?

So that's question 2: How useful is the throttle feature, and is there anything built in which is better? I've a third party external program called "Battle Encoder Shirasé" where you just tell it how much to limit by, and it works flawlessly, but it's another program entirely - is there anything built in?

Question 3: If I let people specify the number of threads, and they pick a number of threads which is higher than what their computer has, I am assuming imagemagick will just default to the max number, and there won't be problems, right? I just want to double check.

Hope someone can clear things up for me, thanks in advance!

**Comment by snibgo** [comment, marked as answer]
> So question 1) When someone wants to specify both a "throttle" and a "thread" number for a magick process, is that the right way to do it? Both specified at the start, one after the other, with a "-limit" before them? And if so, does it matter if thread or throttle goes first? Do both get respected either way?

Yes, to limit both, use "`-limit thread X -limit throttle Y`". The order doesn't matter.

See https://imagemagick.org/script/command-line-options.php#limit . Limits can be set per command or (via policy.xml) per session or per user or system-wide. Limits have no effect on delegates.

> 2: How useful is the throttle feature, and is there anything built in which is better?

Throttling applies only to operations on the pixel cache, so throttling doesn't have much effect.

> Question 3: If I let people specify the number of threads, and they pick a number of threads which is higher than what their computer has, I am assuming imagemagick will just default to the max number, and there won't be problems, right?

Correct. No problems.

**Reply by AV-Gaming** [reply]
What a legend, absolute boss.

Thanks for the quick and accurate response, no beating around the bushes, just direct answers. I'm gonna mark this as correct answer, because it is. Have a great day/week/month.

PS: I just looked at your page, and I didn't think I'd ever find a cooler page than "Fred's ImageMagick Scripts", but yours might come close, and you use batch... man, that's awesome :D I'm porting one of Fred's bash scripts to batch right now!

Thanks again for the reply!
